src/java.base/share/classes/sun/security/ssl/X509Authentication.java

```diff
@@ -204,106 +207,103 @@ private X509PossessionGenerator(String[] keyTypes) {
         public SSLPossession createPossession(HandshakeContext context) {
-            if (context.sslConfig.isClientMode) {
-                for (String keyType : keyTypes) {
-                    SSLPossession poss = createClientPossession(
-                            (ClientHandshakeContext)context, keyType);
-                    if (poss != null) {
-                        return poss;
-                    }
-                }
-            } else {
-                for (String keyType : keyTypes) {
-                    SSLPossession poss = createServerPossession(
-                            (ServerHandshakeContext)context, keyType);
-                    if (poss != null) {
-                        return poss;
-                    }
-                }
-            }
+            return X509Authentication.createPossession(context, keyTypes);
+        }
+    }
 
-            return null;
+    public static SSLPossession createPossession(
+            HandshakeContext context, String[] keyTypes) {
+        if (context.sslConfig.isClientMode) {
+            return createClientPossession(
+                    (ClientHandshakeContext) context, keyTypes);
+        } else {
+            return createServerPossession(
+                    (ServerHandshakeContext) context, keyTypes);
         }
+    }
 
-        // Used by TLS 1.2 and TLS 1.3.
-        private SSLPossession createClientPossession(
-                ClientHandshakeContext chc, String keyType) {
-            X509ExtendedKeyManager km = chc.sslContext.getX509KeyManager();
-            String clientAlias = null;
-            if (chc.conContext.transport instanceof SSLSocketImpl) {
-                clientAlias = km.chooseClientAlias(
-                        new String[] { keyType },
-                        chc.peerSupportedAuthorities == null ? null :
-                                chc.peerSupportedAuthorities.clone(),
-                        (SSLSocket)chc.conContext.transport);
-            } else if (chc.conContext.transport instanceof SSLEngineImpl) {
-                clientAlias = km.chooseEngineClientAlias(
-                        new String[] { keyType },
-                        chc.peerSupportedAuthorities == null ? null :
-                                chc.peerSupportedAuthorities.clone(),
-                        (SSLEngine)chc.conContext.transport);
-            }
+    // Used by TLS 1.2 and TLS 1.3.
+    private static SSLPossession createClientPossession(
+            ClientHandshakeContext chc, String[] keyTypes) {
+        X509ExtendedKeyManager km = chc.sslContext.getX509KeyManager();
+        String clientAlias = null;
+        if (chc.conContext.transport instanceof SSLSocketImpl) {
+            clientAlias = km.chooseClientAlias(
+                    keyTypes,
+                    chc.peerSupportedAuthorities == null ? null :
+                            chc.peerSupportedAuthorities.clone(),
+                    (SSLSocket) chc.conContext.transport);
```

Comment:
Do you want to use pattern matching so as to avoid the cast?

`if (chc.conContext.transport instanceof SSLSocketImpl sslSocket)`